packages/components/src/image_zoom_view/hooks.jsx

```diff
@@ -0,0 +1,180 @@
+import useThrottledCallback from 'nebenan-react-hocs/lib/use_throttled_callback';
+import {
+  getDistanceBetweenPoints,
+  getElementWidth,
+  getMidpoint,
+  getOffsetForMovement, getPixelsFromViewportHeight, getPointDifference,
+  isLengthInThreshold,
+} from './utils';
+import {
+  CONTAINER_WIDTH_CHANGE_RATE,
+  DOUBLE_TAP_THRESHOLD,
+  DOUBLE_TAP_TIMEOUT, DOUBLE_TAP_ZOOM,
+  RESIZE_UPDATE_THRESHOLD,
+} from './constants';
+
+// DO NOT MERGE: EXTRACTION IN PROGRESS
+export const useStateControlledInput = (ref, state) => {
+  useEffect(() => {
+    if (!ref.current) return;
+
+    ref.current.setValue(state, null, { silent: true });
+  }, [state]);
+};
+
+
+export const useDoubleTapZoom = (onAnchorZoom) => {
+  const lastTapRef = useRef({});
+
+  return (point) => {
+    const { point: lastPoint, time: lastTime } = lastTapRef.current;
+    const time = new Date().getTime();
+
+    if (!lastPoint || time - lastTime >= DOUBLE_TAP_TIMEOUT) {
+      lastTapRef.current = { point, time };
+
+      return;
+    }
+
+    if (
+      isLengthInThreshold(lastPoint.x, point.x, DOUBLE_TAP_THRESHOLD)
+      && isLengthInThreshold(lastPoint.y, point.y, DOUBLE_TAP_THRESHOLD)
+    ) {
+      lastPoint.current = {};
+      onAnchorZoom((scale) => scale + DOUBLE_TAP_ZOOM, point);
+    }
+  };
+};
+
+export const useDrag = (onUpdate) => {
+  const dragRef = useRef(null);
+
+  const start = (position, offset) => {
+    const { x: offsetX, y: offsetY } = position;
+    const { top: startTop, left: startLeft } = offset;
+
+    dragRef.current = { offsetX, offsetY, startTop, startLeft };
+  };
+
+  const move = (position) => {
+    const origin = dragRef.current;
+    if (!origin) return;
+
+    const { x: offsetX, y: offsetY } = position;
+
+    onUpdate(
+      getOffsetForMovement(origin, { offsetX, offsetY }),
+    );
+  };
+
+  const stop = () => {
+    dragRef.current = null;
+  };
+
+  return { start, move, stop };
+};
+
+export const usePinchZoom = (onAnchorZoom) => {
+  const pinchZoomRef = useRef(null);
+
+  const start = (pointA, pointB) => {
+    const lastDistance = getDistanceBetweenPoints(pointA, pointB);
+    const lastMidpoint = getMidpoint(pointA, pointB);
+    pinchZoomRef.current = { lastDistance, lastMidpoint };
+  };
+
+  const move = (pointA, pointB) => {
+    const distance = getDistanceBetweenPoints(pointA, pointB);
+    const midpoint = getMidpoint(pointA, pointB);
+
+    const { lastDistance, lastMidpoint } = pinchZoomRef.current;
+    const zoomFactor = distance / lastDistance;
+    const movement = getPointDifference(lastMidpoint, midpoint);
+
+    onAnchorZoom((scale) => scale * zoomFactor, midpoint, movement);
+
+    pinchZoomRef.current = {
+      lastDistance: distance,
+      lastMidpoint: midpoint,
+    };
+  };
+
+  const stop = () => {
+    pinchZoomRef.current = null;
+  };
+
+  return { start, stop, move };
+};
+
+const useWindowResizeEffect = (fn) => {
+  useEffect(() => {
+    fn();
+  }, []);
+
+  useEffect(() => {
```

Comment:
maybe use `useLayoutEffect `with an guard? then we don't need the on-mount effect above